### ISSUE TITLE
error: failed to run custom build command for exonum-proto v1.0.0

### ISSUE THREAD

**Issue by jsoneaday** [michaelvanstraten/actix-jwt-auth-middleware#12]
I don't think this is an issue on actix-jwt but I get this error when building on my docker image. Was wondering if anyone saw this before?

**Comment by michaelvanstraten** [comment]
Do you have the `protoc` compiler installed?

**Comment by jsoneaday** [comment]
Ok I did install that but now I get errors about libsodium, although it appears to be installed. Is there a prereq list of stuff I need to install? Here's the errors I'm seeing.

```sh
#0 237.7   FAIL: pwhash_scrypt
#0 237.7   ============================================================================
#0 237.7   Testsuite summary for libsodium 1.0.17
#0 237.7   ============================================================================
#0 237.7   # TOTAL: 75
#0 237.7   # PASS:  74
#0 237.7   # SKIP:  0
#0 237.7   # XFAIL: 0
#0 237.7   # FAIL:  1
#0 237.7   # XPASS: 0
#0 237.7   # ERROR: 0
#0 237.7   ============================================================================
#0 237.7   See test/default/test-suite.log
#0 237.7   Please report to https://github.com/jedisct1/libsodium/issues
#0 237.7   ============================================================================
#0 237.7   make[4]: Leaving directory '/usr/src/syntaxmakersserver/target/debug/build/exonum_libsodium-sys-0f0c3a9b3a75c0c1/out/source/libsodium-1.0.17/test/default'
#0 237.7   make[3]: Leaving directory '/usr/src/syntaxmakersserver/target/debug/build/exonum_libsodium-sys-0f0c3a9b3a75c0c1/out/source/libsodium-1.0.17/test/default'
#0 237.7   make[2]: Leaving directory '/usr/src/syntaxmakersserver/target/debug/build/exonum_libsodium-sys-0f0c3a9b3a75c0c1/out/source/libsodium-1.0.17/test/default'
#0 237.7   make[1]: Leaving directory '/usr/src/syntaxmakersserver/target/debug/build/exonum_libsodium-sys-0f0c3a9b3a75c0c1/out/source/libsodium-1.0.17/test'
#0 237.7 
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   libtool: warning: '-version-info/-version-number' is ignored for convenience libraries
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   ar: `u' modifier ignored since `D' is the default (see `U')
#0 237.7   ../../build-aux/test-driver: line 107: 55055 Killed                  "$@" > $log_file 2>&1
#0 237.7   make[4]: *** [Makefile:1893: test-suite.log] Error 1
#0 237.7   make[3]: *** [Makefile:2001: check-TESTS] Error 2
#0 237.7   make[2]: *** [Makefile:2593: check-am] Error 2
#0 237.7   make[1]: *** [Makefile:401: check-recursive] Error 1
#0 237.7   make: *** [Makefile:514: check-recursive] Error 1
#0 237.7 
#0 237.7   ', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/exonum_libsodium-sys-0.0.23/build.rs:264:13
#0 237.7   stack backtrace:
#0 237.7      0: rust_begin_unwind
#0 237.7                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
#0 237.7      1: core::panicking::panic_fmt
#0 237.7                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
#0 237.7      2: build_script_build::main
#0 237.7                at ./build.rs:264:13
#0 237.7      3: core::ops::function::FnOnce::call_once
#0 237.7                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/ops/function.rs:250:5
#0 237.7   note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by michaelvanstraten** [comment]
> Is there a prereq list of stuff I need to install?

I don't know what system you are on, I didn't need to install anything. You could check out the [`exonum-crypto`](https://github.com/exonum/exonum) repository for more info, though.

As a quick fix, I would recommend switching to the [`ed25519_dalek`](https://doc.dalek.rs/ed25519_dalek/) crypto implementation.

**Comment by michaelvanstraten** [comment]
Maybe try the following:
```bash
sudo apt-get install -y  build-essential  jq  libsodium-dev  libsnappy-dev  libssl-dev  pkg-config  rocksdb-tools  librocksdb-dev  librocksdb7.8  librust-clang-sys+libloading-dev  librust-clang-sys-dev  libprotobuf-dev  libprotoc-dev  protobuf-compiler
```

**Comment by jsoneaday** [comment]
Yeah that looks good ill give it a shot thanks

**Comment by jsoneaday** [comment]
I must say I'm very surprised but that did not work. I tried switching to Ubuntu 23.04 and your apt install but same error with libsodium on the pwhash_scrypt. Here's my docker file. What do you think any issues?

```docker
FROM rust:1.70 as builder
WORKDIR /usr/src/syntaxmakersserver
COPY . .

RUN curl -OL https://github.com/protocolbuffers/protobuf/releases/download/v24.0/protoc-24.0-linux-x86_64.zip
RUN unzip -o protoc-24.0-linux-x86_64.zip -d /usr/local bin/protoc
RUN unzip -o protoc-24.0-linux-x86_64.zip -d /usr/local include/*
RUN rm -f protoc-24.0-linux-x86_64.zip
RUN CARGO_PROFILE_DEV_BUILD_OVERRIDE_DEBUG=true RUST_BACKTRACE=1 cargo install --debug --path .
 
FROM ubuntu:23.04

RUN apt-get update
RUN apt-get install -y wget
RUN apt-get install -y curl
RUN apt-get install -y unzip
RUN apt-get install -y build-essential  jq  libsodium-dev  libsnappy-dev  libssl-dev  pkg-config  rocksdb-tools  librocksdb-dev  librocksdb7.8  librust-clang-sys+libloading-dev  librust-clang-sys-dev  libprotobuf-dev  libprotoc-dev  protobuf-compiler

COPY --from=builder /usr/src/syntaxmakersserver /usr/local/bin/syntaxmakersserver
WORKDIR /usr/local/bin/syntaxmakersserver
ENTRYPOINT [ "./target/debug/syntaxmakers-server" ]
```

**Comment by jsoneaday** [comment]
According to this, https://github.com/jedisct1/libsodium/discussions/1297, on libsodium github the 1.0.17 version had this bug but is fixed in newer versions. Is there any way to force exonum to use a newer version?
Also are you running on docker. When I run on mac it works. It only fails on docker linux

**Comment by michaelvanstraten** [comment]
Are you on apple Silicon or on x64 intel?

**Comment by jsoneaday** [comment]
This issue is happening on a ubuntu image on docker. On my m1 mac it builds without issue.

**Comment by michaelvanstraten** [comment]
> Is there any way to force exonum to use a newer version?

Only by changing its source, because there is probably some bind-gen magic going on to bind lo lib sodium.

**Comment by michaelvanstraten** [comment]
This should be fixed in [64af5d4](https://github.com/michaelvanstraten/actix-jwt-auth-middleware/commit/64af5d46d0f8734b98cca618e87997e0d0084fa5). It removes the dev dependency on exonum.
